lib/backend/firestore/atomicwrite.go

```diff
@@ -0,0 +1,160 @@
+				if err != nil {
+					if status.Code(err) == codes.NotFound {
+						return trace.Wrap(backend.ErrConditionFailed)
+					}
+					return trace.Wrap(ConvertGRPCError(err))
+				}
+				if isRevisionV2(ca.Condition.Revision) {
+					existingRec, err := newRecordFromDoc(docSnap)
+					if err != nil {
+						return trace.Wrap(err)
+					}
+
+					if existingRec.RevisionV2 != ca.Condition.Revision {
+						return trace.Wrap(backend.ErrConditionFailed)
+					}
+				} else {
+					expectedRev, err := fromRevisionV1(ca.Condition.Revision)
+					if err != nil {
+						return trace.Wrap(backend.ErrConditionFailed)
+					}
+
+					if !docSnap.UpdateTime.Equal(expectedRev) {
+						return trace.Wrap(backend.ErrConditionFailed)
+					}
+				}
+			default:
+				return trace.BadParameter("unexpected condition kind %v in conditional action against key %q", ca.Condition.Kind, ca.Key)
+			}
+		}
+
+		// if we get this far, all conditions hold for this txn context. perform all writes.
+		for _, ca := range condacts {
+			docRef := b.svc.Collection(b.CollectionName).Doc(b.keyToDocumentID(ca.Key))
+
+			switch ca.Action.Kind {
+			case backend.KindNop:
+				// no action to be taken
+			case backend.KindPut:
+				includesPut = true
+				// create shallow copy of item to avoid mutating condacts
+				item := ca.Action.Item
+				item.Key = ca.Key
+				item.Revision = revision
+				newRec := newRecord(item, b.clock)
+				if err := tx.Set(docRef, newRec); err != nil {
+					return trace.Wrap(ConvertGRPCError(err))
+				}
+			case backend.KindDelete:
+				if err := tx.Delete(docRef); err != nil {
+					return trace.Wrap(ConvertGRPCError(err))
+				}
+			default:
+				return trace.BadParameter("unexpected action kind %v in conditional action against key %q", ca.Action.Kind, ca.Key)
+			}
+		}
+
+		return nil
+	}, firestore.MaxAttempts(maxTxnAttempts))
+
+	if err != nil {
+		if status.Code(err) == codes.Aborted {
+			var keys [][]byte
+			for _, ca := range condacts {
+				keys = append(keys, ca.Key)
+			}
+			log.Errorf("AtomicWrite failed, firestore experienced too many txn rollbacks. keys=%s", bytes.Join(keys, []byte(",")))
+			// RunTransaction does not officially document what error is returned if MaxAttempts is exceeded,
+			// but as currently implemented it should simply bubble up the Aborted error from the most recent
+			// failed commit attempt.
+			return "", trace.Errorf("too many attempts during firestore txn for AtomicWrite")
+		}
+
+		return "", trace.Wrap(ConvertGRPCError(err))
+	}
+
+	if n > 2 {
+		// if we retried more than once, txn experienced non-trivial contention and we should warn about it. Infrequent warnings of this kind
+		// are nothing to be concerned about, but high volumes may indicate than an automatic process is creating excessive conflicts.
+		log.Warnf("AtomicWrite retried %d times due to firestore txn rollbacks. Some rollbacks are expected, but persistent rollback warnings may indicate an unhealthy state.", n)
```

Comment:
Perhaps worth adding a metric for backend contention?

Reply:
Chatted with Tim & decided we're going to add a global contention metric once all backend impls are merged and then add it to all of them, so this will be forthcoming in a future PR.